python/tvm/relax/training/setup_trainer.py

```diff
@@ -0,0 +1,121 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=not-callable, unused-argument
+"""Setup Trainer Pass."""
+from typing import List
+
+import tvm
+from tvm.ir.module import IRModule
+
+from ..transform import LegalizeOps, Gradient
+from .loss import Loss
+from .utils import append_loss
+from .optimizer import Optimizer
+from ..struct_info import TensorStructInfo
+from ..analysis import well_formed
+
+
+@tvm.transform.module_pass(opt_level=3, name="SetupTrainer")
```

Comment:
opt level should be 0 here. I guess it was due to the misleading opt level in LegalizeOp, it is now fixed and sorry for that misleading.